pkg/blockbuilder/blockbuilder.go

```diff
@@ -239,3 +240,308 @@ func (b *BlockBuilder) getLagForPartition(ctx context.Context, partition int32)
+	commitRecTs, seenTillOffset, lastBlockEnd, err := unmarshallCommitMeta(lag.Commit.Metadata)
+	if err != nil {
+		// If there is an error in unmarshalling the metadata, treat it as if
+		// we have no commit. There is no reason to stop the cycle for this.
+		level.Warn(logger).Log("msg", "error unmarshalling commit metadata", "err", err, "partition", lag.Partition, "offset", lag.Commit.At, "metadata", lag.Commit.Metadata)
+	}
+
+	if commitRecTs == 0 {
+		// If there was no commit metadata, we use the fallback to replay a set amount of
+		// records because it is non-trivial to peek at the first record in a partition to determine
+		// the range of replay required. Without knowing the range, we might end up trying to consume
+		// a lot of records in a single partition consumption call and end up in an OOM loop.
+		// TODO(codesome): add a test for this.
+		commitRecTs = fallbackMillis
+	}
+
+	return &partitionState{
+		Partition:          lag.Partition,
+		Lag:                lag.Lag,
+		Commit:             lag.Commit,
+		CommitRecTimestamp: time.UnixMilli(commitRecTs),
+		SeenTillOffset:     seenTillOffset,
+		LastBlockEnd:       lastBlockEnd,
+	}
+}
+
+func (b *BlockBuilder) consumePartition(ctx context.Context, state *partitionState, cycleEnd time.Time) error {
+	lagging := cycleEnd.Sub(state.CommitRecTimestamp) > 3*b.cfg.ConsumeInterval/2
+	if !lagging {
+		// Either we did not find a commit offset or we are not lagging behind by
+		// more than 1.5 times the consume interval.
+		// When there is no kafka commit, we play safe and assume seenTillOffset and
+		// lastBlockEnd was 0 to not discard any samples unnecessarily.
+		err := b.consumePartitionCycle(ctx, state, cycleEnd)
+		if err != nil {
+			return fmt.Errorf("consume partition %d: %w", state.Commit.Partition, err)
+		}
+
+		return nil
+	}
+
+	// We are lagging behind. We need to consume the partition in sections.
+	// We iterate through all the ConsumeInterval intervals, starting from the first one after the last commit until the cycleEnd.
+	//cycleEndStartAt := commitRecTime.Truncate(b.cfg.ConsumeInterval).Add(b.cfg.ConsumeInterval + b.cfg.ConsumeIntervalBuffer)
+	cycleEndStartAt, _ := nextCycleEnd(
+		func() time.Time {
+			return state.CommitRecTimestamp
+		},
+		b.cfg.ConsumeInterval,
+		b.cfg.ConsumeIntervalBuffer,
+	)
+	level.Info(b.logger).Log("msg", "partition is lagging behind", "partition", state.Partition, "lag", state.Lag, "cycle_end_start", cycleEndStartAt, "cycle_end", cycleEnd)
+	for ce := cycleEndStartAt; cycleEnd.Sub(ce) >= 0; ce = ce.Add(b.cfg.ConsumeInterval) {
+		// Instead of looking for the commit metadata for each iteration, we use the data returned by consumePartition in the next iteration.
+		err := b.consumePartitionCycle(ctx, state, ce)
+		if err != nil {
+			return fmt.Errorf("consume partition %d: %w", state.Partition, err)
+		}
+
+		// If adding the ConsumeInterval takes it beyond the cycleEnd, we set it to the cycleEnd to not
```

Comment:
i read this three times and I'm still confused. If we `ce = cycleEnd`, then wouldn't the loop action increment it with `b.cfg.ConsumeInterval` and the loop condition fail?

Another way of asking it as: aren't you advancing `ce` by setting it to `cycleEnd`? How would that keep us in the loop?

Reply:
TBF I cannot recall why it's like that either. I've refactored this part now, and, at least the current subset of tests, pass 🙃